### PR TITLE
feat(cli): warn on unrecognized CLI options

### PR DESCRIPTION
**Pull request by sanjanaravikumar-az**
## Summary

- Adds a warning when unrecognized CLI options are passed, since yargs currently silently ignores them
- Handles edge cases: env-injected keys (CDK_* via `.env('CDK')`), undefined defaults, camelCase variants, boolean negations, negativeAliases

This implements a non-breaking warning approach. Unrecognized options are still processed by yargs as before, but users now get feedback that their flag was not recognized.

## Changes

| File | Change |
|------|--------|
| `check-unknown-options.ts` | New utility that detects unrecognized options |
| `cli.ts` | Calls `findUnknownOptions` and emits a warning |
| `check-unknown-options.test.ts` | 10 tests covering all edge cases |

## Test plan

- [x] 10 unit tests passing (known options, aliases, camelCase, negation, env-injected, unknown detection)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/aws-cdk/test/cli/util/check-unknown-options.test.ts [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    const argv = {
+      _: ['deploy'],
+      $0: 'cdk',
+      R: true,
+      rollback: false,
+    };
+    expect(findUnknownOptions(argv)).toEqual([]);
+  });
+
+  test('does not report keys injected by yargs .env("CDK") from environment variables', () => {
```
Comment (mrgrain) [inline packages/aws-cdk/test/cli/util/check-unknown-options.test.ts:79]: but why?

File: packages/aws-cdk/lib/cli/util/check-unknown-options.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ * Detect unrecognized CLI options and emit warnings.
+ *
+ * Yargs does not enable strict option checking by default, so unknown flags
+ * like `--region` (before it was added) are silently swallowed. This function
+ * compares the parsed argv keys against the known global and command options
+ * from the CLI type registry and warns for any that don't match.
+ */
+export function findUnknownOptions(argv: any): string[] {
+  // eslint-disable-next-line @typescript-eslint/no-require-imports
+  const config = require('../cli-type-registry.json');
```
Comment (mrgrain) [inline packages/aws-cdk/lib/cli/util/check-unknown-options.ts]: nit: this doesn't change, so can be outside the function as a "global" variable. In practice this doesn't matter much because `findUnknownOptions` is only called once and `require` is cached, but you should start to thinking about these kind of things.

The same applies to much of the "setup" code later on: sorting global and command options etc.

A good question to ask is: Which parts of the code are actually dependent on the `argv` input? Which parts only depend on the static `cli-type-registry.json`? This will give you an idea how to split it.

File: packages/aws-cdk/lib/cli/util/check-unknown-options.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  if (commandArg) {
+    yargsInternals.add(commandArg);
+  }
+
+  const unknown: string[] = [];
+  for (const key of Object.keys(argv)) {
+    if (argv[key] === undefined) continue;
+    if (yargsInternals.has(key)) continue;
+    if (globalOptions.has(key)) continue;
+    if (commandOptions.has(key)) continue;
```
Comment (mrgrain) [inline packages/aws-cdk/lib/cli/util/check-unknown-options.ts]: can be one `if`

File: packages/aws-cdk/lib/cli/util/check-unknown-options.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    // yargs creates camelCase versions of kebab-case options — skip those
+    const kebab = camelToKebab(key);
+    if (kebab !== key && (globalOptions.has(kebab) || commandOptions.has(kebab))) continue;
+
+    // yargs creates "noFoo" keys for --no-foo boolean negations — skip those
+    if (key.startsWith('no') && key.length > 2 && key[2] === key[2].toUpperCase()) {
+      const positiveKey = key[2].toLowerCase() + key.slice(3);
+      const positiveKebab = camelToKebab(positiveKey);
+      if (globalOptions.has(positiveKey) || commandOptions.has(positiveKey) ||
+          globalOptions.has(positiveKebab) || commandOptions.has(positiveKebab)) continue;
```
Comment (mrgrain) [inline packages/aws-cdk/lib/cli/util/check-unknown-options.ts]: we should enforce this as a linter rule, but imho these lines are hard to parse and read. Prefer always using `{}`.

File: packages/aws-cdk/lib/cli/util/check-unknown-options.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+
+    unknown.push(key);
+  }
+
+  return unknown;
+}
+
+function camelToKebab(str: string): string {
+  return str.replace(/[A-Z]/g, (m) => `-${m.toLowerCase()}`);
+}
```
Comment (mrgrain) [inline packages/aws-cdk/lib/cli/util/check-unknown-options.ts]: This probably happens to work in this case, but is likely not working in the general case. For your code, it would be much easier to go the other way: From kebab to camel.

**Review by mrgrain — DISMISSED** [review]
small feedback.

File: packages/aws-cdk/test/cli/util/check-unknown-options.test.ts (hunk 1 above)
Comment (sanjanaravikumar-az) [reply packages/aws-cdk/test/cli/util/check-unknown-options.test.ts:79]: I dug into it and the test was actually wrong. I ran the actual CDK parser and `yargs` doesn't create `noFoo `keys from `--no-foo` flags, it just sets `foo: false`. The only way a `noFoo` key ends up in argv is if someone sets `CDK_NO_FOO` as an env var (via .`env('CDK')`), which the env prefix check already covers. Removed.

File: packages/aws-cdk/lib/cli/util/check-unknown-options.ts (hunk 2 above)
Comment (sanjanaravikumar-az) [reply packages/aws-cdk/lib/cli/util/check-unknown-options.ts]: That's a good point, thank you! Moved the registry load and global options set to module scope since they never change. Only the command-specific options are computed inside the function now since those depend on which command was invoked.

File: packages/aws-cdk/lib/cli/util/check-unknown-options.ts (hunk 3 above)
Comment (sanjanaravikumar-az) [reply packages/aws-cdk/lib/cli/util/check-unknown-options.ts]: I actually removed this entirely since it's the same issue as the first comment. The negation check was solving a problem that doesn't exist with how yargs works.

File: packages/aws-cdk/lib/cli/util/check-unknown-options.ts (hunk 4 above)
Comment (sanjanaravikumar-az) [reply packages/aws-cdk/lib/cli/util/check-unknown-options.ts]: got it, changed it

File: packages/aws-cdk/lib/cli/util/check-unknown-options.ts (hunk 5 above)
Comment (sanjanaravikumar-az) [reply packages/aws-cdk/lib/cli/util/check-unknown-options.ts]: ah okay, switched it

File: packages/aws-cdk/lib/cli/util/check-unknown-options.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
+// eslint-disable-next-line @typescript-eslint/no-require-imports
+const config = require('../cli-type-registry.json');
+
+/**
+ * Build a set of all known option names for a given option definitions object.
+ * Includes the kebab-case name, its camelCase equivalent, all aliases, and negativeAliases.
+ */
+function collectKnownOptions(optionDefs: Record<string, any>): Set<string> {
```
Comment (iankhou) [inline packages/aws-cdk/lib/cli/util/check-unknown-options.ts:12]: `config.globalOptions` that is passed in contains strings with kebab-case. However, the function signature and JSDoc comments don't indicate that the input must be kebab-case to work correctly. There should be some indication of how to use the function correctly, even though it's not public.

File: packages/aws-cdk/test/cli/util/check-unknown-options.test.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+  test('detects unknown options', () => {
+    const argv = {
+      _: ['bootstrap'],
+      $0: 'cdk',
+      profile: 'my-profile',
+      fakeOption: 'value',
+    };
+    const unknown = findUnknownOptions(argv);
+    expect(unknown).toContain('fakeOption');
```
Comment (iankhou) [inline packages/aws-cdk/test/cli/util/check-unknown-options.test.ts]: We can be stricter with this check. It shouldn't just contain `'fakeOption'`, it should ONLY contain `'fakeOption'`.

File: packages/aws-cdk/test/cli/util/check-unknown-options.test.ts [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    try {
+      const argv = {
+        _: ['deploy'],
+        $0: 'cdk',
+        integAtmospherePool: 'test-pool',
+        totallyFakeOption: 'value',
+      };
+      const unknown = findUnknownOptions(argv);
+      expect(unknown).not.toContain('integAtmospherePool');
+      expect(unknown).toContain('totallyFakeOption');
```
Comment (iankhou) [inline packages/aws-cdk/test/cli/util/check-unknown-options.test.ts]: For this test, I would recommend comparing unknown to an exact object.

**Review by iankhou — CHANGES_REQUESTED** [review]

File: packages/aws-cdk/lib/cli/util/check-unknown-options.ts (hunk 6 above)
Comment (sanjanaravikumar-az) [reply packages/aws-cdk/lib/cli/util/check-unknown-options.ts:12]: added!

File: packages/aws-cdk/test/cli/util/check-unknown-options.test.ts (hunk 7 above)
Comment (sanjanaravikumar-az) [reply packages/aws-cdk/test/cli/util/check-unknown-options.test.ts]: got it, changed to `expect(unknown).toEqual(['fakeOption']);`

File: packages/aws-cdk/test/cli/util/check-unknown-options.test.ts (hunk 8 above)
Comment (sanjanaravikumar-az) [reply packages/aws-cdk/test/cli/util/check-unknown-options.test.ts]: same change, updated to `expect(unknown).toEqual(['totallyFakeOption']);`

**Review by iankhou — APPROVED** [review]
